File: 3.2P/Clock/Clock/Clock.cs
using System;
using System.Data;

namespace Clock
{
    public class Clock
    {
        private Counter _hour;
        private Counter _minute;
        private Counter _second;

        public Clock()
        {
            _hour = new Counter("hour");
            _minute = new Counter("minute");
            _second = new Counter("second");
        }
        
        /// <summary>
        /// method <code> Tick </code>
        /// Increments the 3 Counter objects that make up a Clock (Hour:Minute:Second).
        /// <remarks> I hate the implementation of the nested if statement, I want to make it cleaner. </remarks>
        /// </summary>
        /// <exception cref="InvalidExpressionException"></exception>
        public void Tick()
        {
            if (_second.Ticks == 59)
            {
                _second.Reset();
                if (_minute.Ticks == 59)
                {
                    _minute.Reset();
                    if (_hour.Ticks == 23)
                    {
                        throw new InvalidExpressionException("Program can only handle 24 hours.");
                    }
                    else
                    {
                        _hour.Increment();
                    }
                }
                else
                {
                    _minute.Increment();
                }
            }
            else
            {
                _second.Increment();
            }
        }

        public void Reset()
        {
            _hour.Reset();
            _minute.Reset();
            _second.Reset();
        }

        /// <summary>
        /// <c></c>
        /// property <code> Time </code>
        /// <remarks> Originally used a String.format but Rider complained about it.
        ///           Since found out that apparently in certain situations the interpolation
        ///           is managed at compile time so could have a perf boost
        /// </remarks>
        /// <returns> A string containing the current time of the clock</returns>
        /// </summary>
        public string Time
        {
            get
            {
                return $"{_hour.Ticks:00}:{_minute.Ticks:00}:{_second.Ticks:00}";
            } 
        }
    }
}

File: 10.1P/clock.py
from counter import Counter


class Clock(object):

    def __init__(self):
        self._seconds = Counter("Seconds")
        self._minutes = Counter("Minutes")
        self._hours = Counter("Hours")

    @property
    def time(self) -> str:
        return '{0:02d}:{1:02d}:{2:02d}'.format(self._hours.ticks, self._minutes.ticks, self._seconds.ticks)

    def tick(self) -> None:
        if (self._seconds.ticks != 59):
            self._seconds.increment()
        else:
            self._seconds.reset()
            if (self._minutes.ticks != 59):
                self._minutes.increment()
            else:
                self._minutes.reset()
                if (self._hours.ticks != 23):
                    self._hours.increment()
                else:
                    raise Exception("Out of time! This clock can only go to 24 hours!")
